embassy/src/util/mpsc.rs

```diff
@@ -0,0 +1,905 @@
+                // contain the original message. We could create another UnsafeCell
+                // with the message of Full, but there's no real need.
+            }
+        }
+    }
+}
+
+struct CloseFuture<'ch, M, T, const N: usize>
+where
+    M: Mutex<Data = ()>,
+{
+    sender: Sender<'ch, M, T, N>,
+}
+
+impl<'ch, M, T, const N: usize> Future for CloseFuture<'ch, M, T, N>
+where
+    M: Mutex<Data = ()>,
+{
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        if self.sender.is_closed() {
+            Poll::Ready(())
+        } else {
+            self.sender
+                .channel
+                .get()
+                .set_senders_waker(cx.waker().clone());
+            Poll::Pending
+        }
+    }
+}
+
+impl<'ch, M, T, const N: usize> Drop for Sender<'ch, M, T, N>
+where
+    M: Mutex<Data = ()>,
+{
+    fn drop(&mut self) {
+        self.channel.get().deregister_sender()
+    }
+}
+
+impl<'ch, M, T, const N: usize> Clone for Sender<'ch, M, T, N>
+where
+    M: Mutex<Data = ()>,
+{
+    #[allow(clippy::clone_double_ref)]
+    fn clone(&self) -> Self {
+        self.channel.get().register_sender();
+        Sender {
+            channel: self.channel.clone(),
+        }
+    }
+}
+
+/// An error returned from the [`try_recv`] method.
+///
+/// [`try_recv`]: super::Receiver::try_recv
+#[derive(PartialEq, Eq, Clone, Copy, Debug)]
+pub enum TryRecvError {
+    /// A message could not be received because the channel is empty.
+    Empty,
+
+    /// The message could not be received because the channel is empty and closed.
+    Closed,
+}
+
+/// Error returned by the `Sender`.
+#[derive(Debug)]
+pub struct SendError<T>(pub T);
+
+impl<T> fmt::Display for SendError<T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(fmt, "channel closed")
+    }
+}
+
+/// This enumeration is the list of the possible error outcomes for the
+/// [try_send](super::Sender::try_send) method.
+#[derive(Debug)]
+pub enum TrySendError<T> {
+    /// The data could not be sent on the channel because the channel is
+    /// currently full and sending would require blocking.
+    Full(T),
+
+    /// The receive half of the channel was explicitly closed or has been
+    /// dropped.
+    Closed(T),
+}
+
+impl<T> fmt::Display for TrySendError<T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            fmt,
+            "{}",
+            match self {
+                TrySendError::Full(..) => "no available capacity",
+                TrySendError::Closed(..) => "channel closed",
+            }
+        )
+    }
+}
+
+struct ChannelState<T, const N: usize> {
+    buf: [MaybeUninit<UnsafeCell<T>>; N],
+    read_pos: usize,
+    write_pos: usize,
+    full: bool,
+    closing: bool,
+    closed: bool,
+    receiver_registered: bool,
+    senders_registered: u32,
+    receiver_waker: Option<Waker>,
```

Comment:
These should use WakerRegistration. With the embassy executor, it makes it so wakers are just 1 word instead of 2, and also avoids unnecessary cloning which is a bit more efficient in executors that refcount wakers.

Reply:
Done as per commit 766a3958018b8fd66f34dfe2381678afe683feb7